### PR TITLE
Timelapse microscopy example

### PR DESCRIPTION
**Pull request by droumis**
timelapse_standalone

TODO:
- [x] pin Bokeh 3.6.2 when released and relock. check scalebars

### CODE REVIEW

**Comment by droumis** [comment]
The final app is too wide for the website

image

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 1]
```diff
@@ -0,0 +1,1187 @@
...
+   "metadata": {},
+   "source": [
+    "## App V3: Advanced Viewer with Annotations and Timeseries\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c8d52e0c-eac2-429d-8658-f8ad17c2cfb1",
+   "metadata": {},
+   "source": [
```
Comment (maximlt) [inline timelapse_microscopy/timelapse_microscopy.ipynb:748]: As done before, it'd be nice to show the end state of the app here, so the reader can better approach where this is going.

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 2]
```diff
@@ -0,0 +1,1187 @@
...
+    "In the final version, we'll add annotation capabilities using `HoloNote` and link these annotations to timeseries plots of mean intensity over time.\n"
```
Comment (maximlt) [inline timelapse_microscopy/timelapse_microscopy.ipynb:749]: I missed a little bit of neurosciency narrative here (even if I don't necessarily understand it!), what are we annotating and why?

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 3]
```diff
@@ -0,0 +1,1187 @@
...
+    "        title=f'Frame = {frame}',\n",
+    "        frame_height=da.sizes['height'],\n",
+    "        frame_width=da.sizes['width'],\n",
+    "        cmap='Viridis',\n",
+    "        clim=(0, 20),\n",
+    "        colorbar=True,\n",
+    "        tools=['hover', 'crosshair'],\n",
+    "        toolbar='right',\n",
+    "        apply_hard_bounds=True,\n",
+    "        scalebar=True,\n",
```
Comment (maximlt) [inline timelapse_microscopy/timelapse_microscopy.ipynb:415]: There's no scalebar displayed at all in fact? Bug maybe?

File: test_data/timelapse_microscopy/real_miniscope_uint8.zarr/.zattrs
Comment (maximlt) [inline test_data/timelapse_microscopy/real_miniscope_uint8.zarr/.zattrs:1]: Ok the test data size is slightly more than 4MB. I guess that's fine but we have to be careful with the size of these files (all files in general), as test data files for previous projects were smaller.

File: timelapse_microscopy/anaconda-project.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- datashader>=0.16.3 # auto min pinned 2024-11-21
+- dask>=2024.11.2 # auto min pinned 2024-11-21
+- xarray>=2024.5.0
+- zarr>=2.18.3 # auto min pinned 2024-11-21
+- pyarrow>=18.0.0 # auto min pinned 2024-11-21
+- jupyterlab>=4.3.1 # auto min pinned 2024-11-21
+- holonote>=0.2.1 # 0.2.1: visibility and coloring fixes
+- bokeh>=3.6.0 # 3.6.0: scalebars
+- requests>=2.32.3 # auto min pinned 2024-11-21
+- aiohttp>=3.11.6 # auto min pinned 2024-11-21
```
[the remark this answers is not in the document]
Comment (maximlt) [reply timelapse_microscopy/anaconda-project.yml:44]: Unless these are somehow optional dependencies of another library the example relies on?

**Review by maximlt — CHANGES_REQUESTED** [review]
Another super solid example! I've left suggestions part of the review. My main feedback is that the annotating app still feels experimental (more than what you shared in your previous example), it's a feature that needs users and their feedback to mature. Hopefully this example will help with that!

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 5]
```diff
@@ -0,0 +1,1187 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "75f9be02-0915-4841-bf21-1ad3b69a4045",
+   "metadata": {},
+   "source": [
+    "### Working with Annotations\n",
+    "\n",
+    "How to add new annotations: \n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:1043]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb (hunk 1 above)
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:748]: done

Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:748]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 6]
```diff
@@ -0,0 +1,1187 @@
...
+    "\n",
+    "This workflow demonstrates the display of time-lapse microscopy in neuroscience. Each frame in the image-stack dataset corresponds to a concurrent time sample, typically intended to capture a dynamic process of living cells."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f0de8be5-81d9-4755-b7aa-afbacd5166e3",
+   "metadata": {},
+   "source": [
+    "For example, a dynamic process of interest could be [neural action potentials](https://en.wikipedia.org/wiki/Action_potential), and the data might come from a [miniature microscope](http://miniscope.org/)(**see image in this notebook's header**) that captures the change in fluorescence of special proteins caused by electrochemical fluctuations indicative of neuronal activity. These video-like datasets often contain many more frames in the 'Time' dimension compared to the number of pixels in the height or width of each frame.\n"
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:65]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 7]
```diff
@@ -0,0 +1,1187 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "73cdf8a0-63ad-478d-a6d2-2e8a283ed362",
+   "metadata": {},
+   "source": [
+    "### App Versions\n",
+    "We will build three different visualization approaches to cater to different use cases:\n",
+    "\n",
+    "1. **Basic Viewer:** A one-line application using `hvPlot` for quick visualization of the image stack.\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:76]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 8]
```diff
@@ -0,0 +1,1187 @@
...
+    "2. **Intermediate Viewer with Side Views:** Uses `HoloViews` for additional interactive elements, scalebars, and linked side views, aiding better navigation of the image stack and identification of regions of interest.\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:77]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 9]
```diff
@@ -0,0 +1,1187 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "50d7a758-f56c-4954-85f1-db447e8144cf",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Open the dataset from the local copy\n",
+    "ds = xr.open_dataset(\n",
+    "    DATA_PATH.as_posix(),\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:213]: not required. removed

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 10]
```diff
@@ -0,0 +1,1187 @@
...
+   "id": "0c2d583d-cce0-43f3-803d-ea1d755d16d6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hvplot_app = da.hvplot.image(\n",
+    "    groupby=\"frame\",\n",
+    "    title='hvPlot App',\n",
+    "    cmap='viridis',\n",
+    "    clim=(0, 20),\n",
+    "    aspect=da.sizes['width'] / da.sizes['height'],\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:261]: huzzah

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 11]
```diff
@@ -0,0 +1,1187 @@
...
+    "    widget_location='bottom',\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:262]: I think I'll leave it simpler

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 12]
```diff
@@ -0,0 +1,1187 @@
...
+    ")\n",
+    "hvplot_app"
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:264]: added section

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 13]
```diff
@@ -0,0 +1,1187 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "max_proj = da.max('frame').compute().astype(np.float32)\n",
+    "img_max_proj = max_proj.hvplot.image(\n",
+    "    title='Max Over Time',\n",
+    "    cmap=\"magma\",\n",
+    "    clim=(0,20),\n",
+    "    aspect=da.sizes['width'] / da.sizes['height']\n",
+    ")"
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:290]: now displayed

File: timelapse_microscopy/anaconda-project.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/anaconda-project.yml:44]: required by fsspec in this context. I added a comment

File: timelapse_microscopy/anaconda-project.yml [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+  - "hvplot"
+  - "panel"
+  - "holoviews"
+  - "datashader"
+  - "holonote"
+  categories:
+  - Neuroscience
+  title: "Time-lapse Microscopy"
+  deployments:
+  # - command: notebook
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/anaconda-project.yml:18]: sounds good

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 15]
```diff
@@ -0,0 +1,1187 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "aca6e114-d2dc-4ced-bac1-606fb802b5d9",
+   "metadata": {},
+   "source": [
+    "To provide context over the time dimension, we'll add side views showing the mean intensity over time along the height and width dimensions.\n",
+    "\n",
+    "The right-side view (as if looking at our 3D volume from the right side) will be a frame-by-height view, and the top-side view will be a width-by-frame view. Using `.persist()` (see [relevant Xarray docs](https://docs.xarray.dev/en/stable/user-guide/dask.html#using-dask-with-xarray) allows us to cache the results of the mean calculations, reducing recomputation and improving performance. `Rasterizing` these views helps to limit the amount of data sent to the browser, ensuring efficient rendering (see [relevant HoloView's docs](https://holoviews.org/user_guide/Large_Data.html)). We will also set `axiswise=True` to prevent the side views from adjusting to the range of the main view, as we want them to be stable references of the full range on their respective axes."
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:561]: nice, thanks for the validation and benchmarking. I added a paragraph to drive the point

[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:561]: I think `.compute()` loads the dataset into memory as a single object, but `.persist()` keeps the lazy chunking in place. I added some explainer content

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 16]
```diff
@@ -0,0 +1,1187 @@
...
+   "metadata": {},
+   "source": [
+    "### Working with Annotations\n",
+    "\n",
+    "How to add new annotations: \n",
+    "1. Ensure that 'Box select (x-axis)' is selected in the Bokeh Toolbar on the right of the main frame plot.\n",
+    "2. Ensure that the '+' button is selected in the Annotation widgets on the very left.\n",
+    "3. Select a spatial range by clicking and dragging on the main frame plot.\n",
+    "4. Enter the name of the group that you want the annotation to be added to in the 'type' text box.\n",
+    "5. Edit or delete annotations using the provided widgets and table.\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:1048]: fixed

File: timelapse_microscopy/timelapse_microscopy.ipynb (hunk 3 above)
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:415]: waiting for [Bokeh 3.6.2](https://github.com/bokeh/bokeh/issues/14153)

Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:415]: reminder to self to relock when bokeh pin updated

File: timelapse_microscopy/timelapse_microscopy.ipynb [hunk 17]
```diff
@@ -0,0 +1,1187 @@
...
+    "    accent=\"#30023f\"\n",
+    ").servable()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "585095bc-2402-4549-ac17-a1bf55e1faa2",
+   "metadata": {},
+   "source": [
+    "## Copyable V3 App Code\n",
```
[the remark this answers is not in the document]
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:1198]: done

File: timelapse_microscopy/timelapse_microscopy.ipynb (hunk 3 above)
Comment (droumis) [reply timelapse_microscopy/timelapse_microscopy.ipynb:462]: added as todo in descr to unblock rebase on main

**Review by maximlt — COMMENTED** [review]
Thanks for the updates @droumis !

One suggestion I want to circle back to is https://github.com/holoviz-topics/examples/pull/452#discussion_r1860516344.

> Don't you think it'd be easier for the reader/user of this code if it wasn't suddenly pure HoloViews but hvplot(...).opts(...) (if that's possible of course)? I think that would be my preference.

Your answer:

> Especially since it's building up in complexity with holoviews through the v2, v3 apps, IMO it's better to just switch to holoviews at the start of v2 rather than mix hvplot + opts and then holoviews elsewhere in the same app.

I'm fine with your answer of not updating the code. However, I'd like to see at least some text explanation linking hvPlot and HoloViews. Composition is a strong value of HoloViz so my suggestion is to mention it when we can. I'm suggesting that as, from an outsider's perspective, hvPlot could be seen as a sort of dead-end used only to make simple plots but that cannot be used to help creating more complex HoloViews/Panel dashboards. I'm also unsure readers will realize how related hvPlot and HoloViews are.

Besides that, LGTM!

**Review by maximlt — APPROVED** [review]

**Comment by droumis** [comment]
> I'm fine with your answer of not updating the code. However, I'd like to see at least some text explanation linking hvPlot and HoloViews. Composition is a strong value of HoloViz so my suggestion is to mention it when we can. I'm suggesting that as, from an outsider's perspective, hvPlot could be seen as a sort of dead-end used only to make simple plots but that cannot be used to help creating more complex HoloViews/Panel dashboards. I'm also unsure readers will realize how related hvPlot and HoloViews are.

added

**Pull request merged**